---
title: "GSEA on RNA-seq data"
output: html_notebook
---

Zoë Steier

Use bulk RNA-seq data to find DE between TCR transgenic mice with different TCR signal strengths.
Run GSEA on DESeq2 results to find gene sets and pathway differences between TCR transgenic mice with different levels of self-reactivity.

Gene sets to include:
* H1 hallmarks
* C2 canonical pathways
* C5 GO biological processes
* C5 GO molecular functions
* C7 immunologic signatures.

# Load packages
```{r Load required packages}
library(clusterProfiler)
library(tidyverse)
library(rlang)
library(stringr)
library(purrr)
library(org.Mm.eg.db)
library(org.Hs.eg.db)

library(biomaRt)
library(stringr)
library(fgsea)
library(parallel)
library(openxlsx)

library(BiocParallel)
register(MulticoreParam(workers = 5))
```

# Load data

```{r Load DE results for GSEA}
WaldStats <- read_csv("../DE_RNAseq/DE_results/DESeq2_TCRsignal_CD8_all_WaldStat.csv")
dim(WaldStats)
```


```{r Collect list of genes to remove from DE result}
# Remove intentionally manipulated genes like all TCR genes
patterns <- c("Traj.*", "Trav.*", "Trbc.*", "Trbd.*", "Trbv.*", "Trbj.*", "Trdv.*", "Trdj.*", "Trdc.*", "Trdd.*", "Trgj.*", "Trgv.*", "Trac.*", # TCR genes
              "H2-.*", "Tap1", "Tap2", "Tapbp", "Tapbpl", "Psmb8", "Psmb9", # H2 genes
              "Rag1", "Rag2" # Rag genes
              )

WaldStats_filt <- WaldStats %>%
  filter(!(grepl(str_c(patterns, collapse="|"), Gene)))

```

# Prepare data for fgsea
## Get gene names in correct form

```{r Convert mouse gene symbols to human for comparison with ImmSig databases}
# Map mouse symbols to human symbols
martMouse <- biomaRt::useDataset("mmusculus_gene_ensembl", mart=useMart("ensembl"))
orgdb_Species <- org.Mm.eg.db


bmMouse <- getBM(attributes=c("ensembl_gene_id","hsapiens_homolog_associated_gene_name"),mart=martMouse)
    
# Note: biomart will not allow me to retrieve all three attribute (gene id, gene symbol, human homolog)
# in one call of getBM, so I do these separately and merge.
bmSymbol <- getBM(attributes=c("mgi_symbol","ensembl_gene_id"),mart=martMouse)

tbMouseSymbol_to_MouseENSEMBL <-  bmSymbol %>% distinct() %>% as_tibble() %>% na_if("") %>% na.omit() %>% 
  dplyr::select(Symbol=mgi_symbol,ENSEMBL=ensembl_gene_id)

tbMouseENSEMBL_to_HumanSymbol<- bmMouse %>% distinct() %>% as_tibble() %>% na_if("") %>% na.omit() %>% 
  dplyr::select(ENSEMBL=ensembl_gene_id,Human_Symbol=hsapiens_homolog_associated_gene_name)

tbMouseSymbol_to_HumanSymbol <- dplyr::inner_join(tbMouseSymbol_to_MouseENSEMBL , tbMouseENSEMBL_to_HumanSymbol,by="ENSEMBL") %>% 
  dplyr::select(Symbol,Human_Symbol)

tbMapping <- tbMouseSymbol_to_HumanSymbol
```

```{r Get signature collections for fgsea}

# Gene sets can be downloaded from MSigDB
gmtHallmark <- fgsea::gmtPathways("/data/yosef2/users/zsteier/references/Signatures/h.all.v7.0.symbols.gmt")
gmtC2 <- fgsea::gmtPathways("/data/yosef2/users/david.detomaso/Signatures/MSIGDB/c2.cp.v6.1.symbols.gmt")
gmtC5 <- fgsea::gmtPathways("/data/yosef2/users/david.detomaso/Signatures/MSIGDB/c5.all.v4.0.symbols.gmt")
gmtC7<- fgsea::gmtPathways("/data/yosef2/users/david.detomaso/Signatures/MSIGDB/C7_IMMSIG_ALL.gmt")

```





# Run fgsea
## Run multiple tests
```{r Run multiple fgsea tests}
gene_sets <- list("C2" = gmtC2, "C5" = gmtC5, "C7" = gmtC7, "Hallmark" = gmtHallmark)

for (gene_set_num in seq(length(gene_sets))) {
  gene_set_name <- names(gene_sets)[gene_set_num]

  liResults <- list()
  
  for (i in seq(length(colnames(WaldStats_filt)) - 1)) {
    comparison_name <- str_replace(colnames(WaldStats_filt)[i + 1], "_stat", "")

    test_stat <- WaldStats_filt[, c(1, 1 + i)]
    colnames(test_stat)[2] <- "stat"
    
    test_stat <- test_stat %>%
      group_by(Gene) %>%
      summarize(mean_stat = mean(stat)) %>%
      na.omit() %>%
      left_join(tbMapping, by = c("Gene" = "Symbol")) %>% # convert to human names
      dplyr::select(Human_Symbol, mean_stat) %>% # keep just human gene name
      na.omit() %>% # remove rows with no human gene name, resulting in 16,223 genes
      group_by(Human_Symbol) %>%
      summarize(mean_stat_hu = mean(mean_stat)) %>% # take mean of human gene names, now 14977 genes
      arrange(desc(mean_stat_hu)) %>% # sort from highest to lowest Wald Stat
      deframe()
    
    res_fgsea <- fgsea(pathways=gene_sets[[gene_set_num]],stats=test_stat,nperm=1000,nproc=5)
    tbGSEA <- res_fgsea %>% 
      as_tibble() %>% 
      arrange(desc(NES)) %>% # NES – enrichment score normalized to mean enrichment of random samples of the same size
      dplyr::select(pathway, padj, NES)
    colnames(tbGSEA)[2:3] <- paste(comparison_name,colnames(tbGSEA)[2:3],sep="_") # don't add to pathway column
    
    liResults[[i]] <- (tbGSEA)
  }
  
  # Merge results
  MergeGSEA <- function(tbLeft,tbRight){
    left_join(tbLeft,tbRight,by=c("pathway"))
  }
  
  tbGSEAMerged <- Reduce(MergeGSEA,liResults)
  tbGSEAMerged <- tbGSEAMerged %>%
    dplyr::select(pathway, everything()) # Results for one gene set for all tests
  
  
  write_csv(tbGSEAMerged, str_c("fgsea_results/", gene_set_name, "_results.csv"))
  # Save one output csv per gene set
  
}

getwd()
```

```{r View leading edge genes}
leadedge <- res_fgsea %>% as_tibble() %>%
  dplyr::select(leadingEdge)

leadedge[[1]][1]
```

```{r View GSEA results}
res_fgsea[1:5, 1:8]
```


## Find leading edge genes for selected tests

We are interested in C5 gene sets.

For OT-1 at all developmental stages, we would like to know the leading edge genes for:

1.     STRUCTURAL_CONSTITUENT_OF_RIBOSOME

2.     TRANSLATION

3.     STRUCTURAL_MOLECULE_ACTIVITY

4.     CHROMOSOME_PART

5.     RNA_PROCESSING

 

For TG6 at all developmental stages, we would like to know the leading edge genes for:

1.     VOLTAGE_GATED_CHANNEL_ACTIVITY

2.     VOLTAGE_GATED_CATION_CHANNEL_ACTIVITY

3.     KINASE_BINDING

4.     VOLTAGE_GATED_POTASSIUM_CHANNEL ACTIVITY

5.     VOLTAGE_GATED_CALCIUM_CHANNEL_COMPLEX

6.     CALCIUM_CHANNEL_ACTIVITY

 
Note: the same tests can be used for OT1 vs TG6 and the opposite because leading edge genes are calculated at the maximum deviation from zero (i.e. whether the enrichment score is positive or negative).

```{r For OT1 vs TG6 at all stages}
# which DE tests to use
WaldStats_selected <- WaldStats_filt[, c(1, 4, 8, 14)] # OT1 vs TG6 at immature, semi-mature, mature

gene_sets <- list("C5" = gmtC5)

for (gene_set_num in seq(length(gene_sets))) {
  gene_set_name <- names(gene_sets)[gene_set_num]

  for (i in seq(length(colnames(WaldStats_selected)) - 1)) {
    comparison_name <- str_replace(colnames(WaldStats_selected)[i + 1], "_stat", "")

    test_stat <- WaldStats_selected[, c(1, 1 + i)]
    

    # Convert to human gene names from mouse
    colnames(test_stat)[2] <- "stat"
    
    test_stat <- test_stat %>%
      group_by(Gene) %>%
      summarize(mean_stat = mean(stat)) %>%
      na.omit() %>%
      left_join(tbMapping, by = c("Gene" = "Symbol")) %>% # convert to human names
      dplyr::select(Human_Symbol, mean_stat) %>% # keep just human gene name
      na.omit() %>% # remove rows with no human gene name, resulting in 16,223 genes
      group_by(Human_Symbol) %>%
      summarize(mean_stat_hu = mean(mean_stat)) %>% # take mean of human gene names, now 14977 genes
      arrange(desc(mean_stat_hu)) %>% # sort from highest to lowest Wald Stat
      deframe()

    
    # Run gsea test
    res_fgsea <- fgsea(pathways=gmtC5,stats=test_stat,nperm=1000,nproc=5)

    # Extract leading edge genes for each gene set
    leadedge <- res_fgsea %>% as_tibble() %>%
      dplyr::select(leadingEdge)

    # Find pathways of interest
    OT1_pathways = c("STRUCTURAL_CONSTITUENT_OF_RIBOSOME", "TRANSLATION", "STRUCTURAL_MOLECULE_ACTIVITY", "CHROMOSOMAL_PART","RNA_PROCESSING")
    TG6_pathways = c("VOLTAGE_GATED_CHANNEL_ACTIVITY","VOLTAGE_GATED_CATION_CHANNEL_ACTIVITY","KINASE_BINDING","VOLTAGE_GATED_POTASSIUM_CHANNEL_ACTIVITY","VOLTAGE_GATED_CALCIUM_CHANNEL_COMPLEX","CALCIUM_CHANNEL_ACTIVITY")
    all_pathways <- c(OT1_pathways, TG6_pathways)
    
    
    leading_genes <- (res_fgsea) %>%
      filter(pathway %in% all_pathways)
    
    # Convert list into string
    leading_genes$leadingEdgestr = ""
    for (i in seq(dim(leading_genes)[1])) {
      leading_genes[i, "leadingEdgestr"] <- (str_c(unlist(leading_genes[i, "leadingEdge"]), collapse = ", "))
    }
        
    leading_str <- leading_genes %>%
      dplyr::select(-c("leadingEdge"))
    
    write_csv(leading_str, str_c("fgsea_results/", comparison_name, "_leadingEdgeGenes.csv"))

  }
}

```

